sbi/utils/sbiutils.py

```diff
@@ -141,40 +142,43 @@ def sample_posterior_within_prior(
     samples = []
     num_remaining = num_samples
     num_sampled_total = 0
-    tstart = time.time()
-    time_over = time.time() - tstart > (patience * 60)
 
-    while num_remaining > 0 and not time_over:
-
-        # XXX: we need this reshape here because posterior_nn.sample sometimes return
-        # leading singleton dimension instead of (num_samples), e.g., (1, 10000, 4)
-        # instead of (10000, 4). and this can't be handled by MultipleIndependent. issue #141
-        sample = posterior_nn.sample(num_remaining, context=x).reshape(
-            num_remaining, -1
-        )
-        num_sampled_total += num_remaining
-
-        is_within_prior = torch.isfinite(prior.log_prob(sample))
-        num_in_prior = is_within_prior.sum().item()
-
-        if num_in_prior > 0:
-            samples.append(sample[is_within_prior,].reshape(num_in_prior, -1))
-            num_remaining -= num_in_prior
-
-        # update timer
-        time_over = time.time() - tstart > (patience * 60)
+    # we might not always want a progressbar for sampling. E.g. we want to display that,
+    # after each round, we also draw samples just for logging. Thus, the progressbar
+    # can be turned off with the verbose argument.
+    if not show_progressbar:
+        pbar = no_tqdm()
+    else:
+        pbar = tqdm(total=num_remaining)
+        desc = "Drawing {0} posterior samples".format(num_remaining)
+        if type(show_progressbar) == str:
+            desc += show_progressbar
+        pbar.set_description(desc)
+
+    with pbar:
+        while num_remaining > 0:
```

Comment:
If leakage is very high, then this would be almost an infinite loop, no? Maybe we should add a running estimate of leakage and warn while sampling once it is, say, above `0.99`?

Reply:
Thanks, yes!